### PR TITLE
replace across+1inch with lifi, simplify fee system, rework smart transfers

### PR DESCRIPTION
**Pull request by alexhooketh**
title

### CODE REVIEW

File: contracts/src/UntronFees.sol [hunk 1]
```diff
@@ -1,82 +1,59 @@
...
         internal
         view
-        returns (uint256 value, uint256 _relayerFee)
+        returns (uint256 value, uint256 fee)
     {
         // convert size into USDT L2 based on the rate
         uint256 out = (size * rate / bp);
+        // subtract fixed fulfiller fee from the output value
+        if (includeFulfillerFee) {
+            fee += (value - (value - fulfillerFee));
```
Comment (ziemen4) [inline contracts/src/UntronFees.sol:45]: Did you mean `out` instead of `value`?

File: contracts/src/UntronFees.sol [hunk 2]
```diff
@@ -1,82 +1,59 @@
...
+            value -= fee;
+        }
         // if the relayer fee is included, subtract it from the converted size
         if (includeRelayerFee) {
             // subtract relayer fee from the converted size
             value = out * (bp - relayerFee) / bp;
```
Comment (ziemen4) [inline contracts/src/UntronFees.sol:51]: Mutating this variable again seems rather confusing. Would suggest just using a new variable for this scope instead

File: contracts/src/UntronTransfers.sol [hunk 3]
```diff
@@ -46,110 +37,34 @@ abstract contract UntronTransfers is
...
+        if (transfer.directTransfer) {
+            internalTransfer(usdt, abi.decode(transfer.data, (address)), amount);
+            success = true;
         } else {
-            // otherwise, perform a cross-chain transfer through the Across bridge.
-            // see https://docs.across.to/use-cases/instant-bridging-in-your-application/bridge-integration-guide for reference
+            // otherwise, perform a cross-chain transfer through the LiFi Diamond contract.
+            // see https://docs.lifi.io/developers/api-references/diamond/ for reference
 
             // approving the token to the spoke pool
```
Comment (ziemen4) [inline contracts/src/UntronTransfers.sol:64]: Fix comment

File: contracts/src/interfaces/IUntronTransfers.sol [hunk 4]
```diff
@@ -7,30 +7,12 @@ pragma solidity ^0.8.20;
...
-        // minimum amount of output tokens to receive per 1 USDT L2.
-        uint256 minOutputPerUSDT;
-        // whether the minimum amount of output tokens is fixed.
-        // if true, the order creator will receive exactly minOutputPerUSDT * amount of output tokens.
-        // if false, the order creator will receive at least minOutputPerUSDT * amount of output tokens.
-        bool fixedOutput;
-        // data for the swap. Not used if doSwap is false.
-        bytes swapData;
+        bool directTransfer;
+        bytes data;
```
Comment (ziemen4) [inline contracts/src/interfaces/IUntronTransfers.sol:11]: What exactly do we encode here? Maybe worth commenting

File: contracts/src/UntronCore.sol [hunk 5]
```diff
@@ -224,15 +224,13 @@ contract UntronCore is Initializable, OwnableUpgradeable, UntronTransfers, Untro
...
     /// @return amount The amount of USDT L2 that the fulfiller will have to send
-    /// @return fee The fee for the fulfiller
-    function _getAmountAndFee(Order memory order) internal view returns (uint256 amount, uint256 fee) {
-        // calculate the fulfiller fee given the order details
-        fee = calculateFee(order.transfer.doSwap, order.transfer.chainId);
+    /// @return _fulfillerFee The fee for the fulfiller
+    function _getAmountAndFee(Order memory order) internal view returns (uint256 amount, uint256 _fulfillerFee) {
         // calculate the amount of USDT L2 that the fulfiller will have to send
-        (amount,) = conversion(order.size, order.rate, fee, true);
+        (amount, _fulfillerFee) = conversion(order.size, order.rate, false, true);
```
Comment (ziemen4) [inline contracts/src/UntronCore.sol:233]: Idk if its ok but here behaviour was changed.
Before, `includeRelayerFee` was true and now it is false

File: contracts/src/UntronCore.sol [hunk 6]
```diff
@@ -365,14 +359,10 @@ contract UntronCore is Initializable, OwnableUpgradeable, UntronTransfers, Untro
                 closedOrders[i].inflow < _orders[orderId].size ? closedOrders[i].inflow : _orders[orderId].size;
 
             // calculate the amount the order creator/fulfiller will receive and fee for the protocol
-            (uint256 amount, uint256 fee) = conversion(minInflow, _orders[orderId].rate, 0, true);
+            (uint256 amount, uint256 fee) = conversion(minInflow, _orders[orderId].rate, true, true);
```
Comment (ziemen4) [inline contracts/src/UntronCore.sol:362]: Here before we were having `fixedFee = 0` and now we are including fulfiller fee

Comment (ziemen4) [reply contracts/src/UntronCore.sol:362]: Actually I think that when closing we dont pay the fulfiller right? Just the relayer. Fulfiller gets paid by sending less funds

File: contracts/src/UntronFees.sol (hunk 1 above)
Comment (alexhooketh) [reply contracts/src/UntronFees.sol:45]: yes

File: contracts/src/UntronCore.sol (hunk 5 above)
Comment (alexhooketh) [reply contracts/src/UntronCore.sol:233]: it wasn't true, the mapping was "fulfiller fee, includeRelayerFee" and now it's "includeFulfillerFee, includeRelayerFee

File: contracts/src/UntronCore.sol (hunk 6 above)
Comment (alexhooketh) [reply contracts/src/UntronCore.sol:362]: ok the way i wanted to build this is order.isFulfilled ? false : true

**Comment by alexhooketh** [comment]
we will need to migrate tests to this btw

File: contracts/src/UntronCore.sol (hunk 6 above)
Comment (ziemen4) [reply contracts/src/UntronCore.sol]: Oh true it depends on if it was fulfilled or not

File: contracts/src/UntronCore.sol (hunk 5 above)
Comment (ziemen4) [reply contracts/src/UntronCore.sol:233]: `function conversion(uint256 size, uint256 rate, bool includeRelayerFee, bool includeFulfillerFee)`
Maybe params should be swapped?

File: contracts/src/UntronFees.sol [hunk 7]
```diff
@@ -1,82 +1,59 @@
...
         view
-        returns (uint256 value, uint256 _relayerFee)
+        returns (uint256 value, uint256 fee)
     {
         // convert size into USDT L2 based on the rate
         uint256 out = (size * rate / bp);
+        // subtract fixed fulfiller fee from the output value
+        if (includeFulfillerFee) {
+            fee += (out - (out - fulfillerFee));
+            out -= fee;
```
Comment (ziemen4) [inline contracts/src/UntronFees.sol:46]: Wait is the relayer fee calculated relative to `out` or to `out - fulfillerFee`?

Comment (alexhooketh) [reply contracts/src/UntronFees.sol:46]: the way i designed it, relative to `out - fulfillerFee`

do you think it's a wrong design?

**Pull request merged**